### PR TITLE
hypr audit source

### PR DESCRIPTION
**Pull request by bazsi**
This pull request merges in @danelder 's work on the Hypr audit trail API into the syslog-ng source tree. It changes the user interface and implementation significantly, but changes were documented in the README.md file in the source tree.

This PR contains the following items:
  * hypr-audit-trail() & hypr-app-audit-trail() source: see the README.md file in python-modules/syslogng/modules/hypr directory

The testing for hypr relies on a mocked HTTP backend, which is obviously not very thorough but does validate that the code more-or-less functions. I am unable to try the hypr source with an actual API endpoint as I lack the credentials.

Depends on #4455, #4459

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by MrAnno** [comment]
I've rebased this PR onto the current master to check the CI and make the review easier.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by bazsi** [comment]
rebased against current master.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by alltilla** [comment]
After talking about this IRL, we concluded that we should implement the binding improvements and the correct ack tracking here, before merging this PR. This won't fit to this release, so postponing this to 4.2.0.

**Comment by alltilla** [comment]
* Rebased to master
* Added typed option support
* Added `flags(no-parse)` support
* Added NV pair extraction

TODO:
* Proper bookmarking

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by alltilla** [comment]
I think this is review ready now.
I haven't tried it with a real setup, so I am going to do that next.

@MrAnno Could you take a look at this some time? Thanks! :)

**Comment by alltilla** [comment]
Well, there is no publicly available instance of Hypr, so I have mimicked it with the example response of https://apidocs.hypr.com/#a766b64a-b303-40bb-a02f-2ca9ab2a455e. This is far from ideal :/

@danelder Could you test this in your env perhaps? Thanks!

File: modules/python-modules/syslogng/modules/hypr/__init__.py [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+    page_size = 100
+    initial_hours = 4
+
+    # state
+    start_time = 0
+    end_time = 0
+
+    #
+    logger = logging.getLogger('hypr')
+    persist = None
```
Comment (MrAnno) [inline modules/python-modules/syslogng/modules/hypr/__init__.py:74]: These are class attributes, which are then shadowed by instance attributes.
Creating a constructor would probably be cleaner.

File: modules/python-modules/syslogng/modules/hypr/__init__.py [hunk 2]
```diff
@@ -123,22 +124,21 @@ def init(self, options):
...
+                self.max_performance = options["max_performance"]
 
         # Default time to go back is 4 hours
         initial_hours = 4
         if "initial_hours" in options:
 
             # If r is set as an initial_hours value, ignore persistence
-            if "r" in options["initial_hours"]:
-                self.logger.info("Disabling persistence due to special initial_hours setting (%s)", options["initial_hours"])
+            if "r" in str(options["initial_hours"]):
```
Comment (MrAnno) [inline modules/python-modules/syslogng/modules/hypr/__init__.py]: We use the str() on initial_hours. Shouldn't we leave the quotes on the SCL side?

**Review by MrAnno — APPROVED** [review]

**Comment by MrAnno** [comment]
LGTM (I haven't tested it).

File: modules/python-modules/syslogng/modules/hypr/__init__.py (hunk 2 above)
Comment (alltilla) [reply modules/python-modules/syslogng/modules/hypr/__init__.py]: At the final version `initial_hours` is not used as an int, and not casted to str and parsed, etc.

File: modules/python-modules/syslogng/modules/hypr/__init__.py (hunk 1 above)
Comment (alltilla) [reply modules/python-modules/syslogng/modules/hypr/__init__.py:74]: Yes, this is ugly, but works. Fortunately `logs` is not created on class level.

**Pull request merged**